### ISSUE TITLE
Hyperparameter tuning on streaming_convnets

### ISSUE THREAD

**Issue by sakares** [facebookresearch/wav2letter#551]
I have successfully reproduced the [streaming_convnets on Librispeech](https://github.com/facebookresearch/wav2letter/tree/master/recipes/models/streaming_convnets/librispeech) and try to adopt these papers to Thai language voice dataset.

Turned out my Thai language model did not yield a good score for train/dev WER for training around 8 hours with 170 epochs.

So, my following questions are:
- What's the suggested number of hours for training datasets? (i.e. 10 hours+)
- Do you have any scheme for optimizing hyperparameter? such as RandomSearch/GridSearch or something
- Not sure if wav2letter have something like [pytorch lr_finder](https://github.com/davidtvs/pytorch-lr-finder) for finding the optimal lr 

Thank you!

**Comment by vineelpratap** [comment]
> What's the suggested number of hours for training datasets? (i.e. 10 hours+)
Hi, 8 hours is very small. We usually need atleast a thousand hours to train good model. Also, for Thai, you should be looking at Letter Error Rate and nor WER as the Thai language doesn't have a good notion of word. Also, did you train model using the data on any other system to know what's the expected number ?

> Do you have any scheme for optimizing hyperparameter? such as RandomSearch/GridSearch or something
> Not sure if wav2letter have something like pytorch lr_finder for finding the optimal lr

No, we currently don't have anything of this sort. Just a grid search on learning rate should be fine.

**Comment by sakares** [comment]
Hi @vineelpratap Thanks for your reply

>  Also, for Thai, you should be looking at Letter Error Rate and nor WER as the Thai language doesn't have a good notion of word.

So, I should change the variable in **.cfg** files to `--decodertype=tkn` which suite to letter training, right? or do you have any suggested recipe for this case? please advise.

> Also, did you train model using the data on any other system to know what's the expected number?

Not yet, I did not try Kaldi or other methods yet. Also, for the dataset, I have carefully selected and scraped the audio source in the narrative style which I think It's closed to Librispeech speaking speed and style.

**Comment by lunixbochs** [comment]
On your dataset size, you should try a few things:
1. Find or make more data, I don't think you'll have any remotely usable results with less than 100 hours. Once you have _any_ model, you can try to use it to align or pseudo label a larger corpus.
2. Try transfer learning. The simple way to do that is just start with the weights of a working English model and zero the token layer. For better results maybe pick a language that's similar to Thai that has more data available. This issue talks about Transfer Learning with wav2letter: https://github.com/facebookresearch/wav2letter/issues/507#issuecomment-580809478
3. Augment your data. Are you training with SpecAugment? The only way to do that in wav2letter master is to add a SAUG layer, which Streaming Convnets doesn't have. Other methods include speed perturbation (generate identical copies of the files which are slightly slowed down or sped up).
4. Try a different model. Perhaps conv_glu (wsjarch / wsj, not librispeech, because you have so little data), or perhaps the early seq2seq_tds model: https://github.com/facebookresearch/wav2letter/tree/master/recipes/models/seq2seq_tds

I found these two whitepapers about Thai recognition:
- http://www.cs.cmu.edu/~awb/papers/icassp2005/0100857.pdf
- https://www.academia.edu/20451707/Tile_name_Thai_Speech_Corpus_for_Speech_Recognition

**Comment by sakares** [comment]
Hi @lunixbochs 
Thanks for pointing out. I also forgot about the transfer learning option. Currently, I am about to gain more hours dataset and implementing the pseudo label strategy too. 

I close this issue and would update later on. Thanks you all for the valuable comments.

**Comment by tlikhomanenko** [comment]
Hi @sakares 

> So, I should change the variable in **.cfg** files to `--decodertype=tkn` which suite to letter training, right? or do you have any suggested recipe for this case? please advise.

Flag `--decodertype=tkn` is a flag only for the beam search decoder, and for Thai language you should use it `--decodertype=tkn` with Language model trained on token level (not word level).

For the training you can just prepare transcription in the list file with spaces between each token and create fake lexicon where you map each token to itself. In this case during training both WER and LER will be equal and will show exactly LER (letter error rate) which you need to measure.
